base/skipoftype.jl

```diff
@@ -0,0 +1,143 @@
+function mapreduce(f, op, itr::SkipOfType{T, <:AbstractArray}) where T
+    _mapreduce(f, op, IndexStyle(itr.x), eltype(itr.x) >: T ? itr : itr.x)
+end
+
+function _mapreduce(
+    f, op, ::IndexLinear, itr::SkipOfType{T, <:AbstractArray}
+) where T
+    A = itr.x
+    local ai
+    inds = LinearIndices(A)
+    i = first(inds)
+    ilast = last(inds)
+    while i <= ilast
+        @inbounds ai = A[i]
+        ai isa T  || break
+        i += 1
+    end
+    i > ilast && return mapreduce_empty(f, op, eltype(itr))
+    a1 = ai
+    i += 1
+    while i <= ilast
+        @inbounds ai = A[i]
+        ai isa T || break
+        i += 1
+    end
+    i > ilast && return mapreduce_first(f, op, a1)
+    # We know A contains at least two non-nothing entries: the result cannot be nothing
+    something(mapreduce_impl(f, op, itr, first(inds), last(inds)))
+end
+
+_mapreduce(f, op, ::IndexCartesian, itr::SkipOfType) = mapfoldl(f, op, itr)
+
+mapreduce_impl(f, op, A::SkipOfType, ifirst::Integer, ilast::Integer) =
+    mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
+
+# Returns nothing when the input contains only nothing values, and Some(x) otherwise
+@noinline function mapreduce_impl(
+    f, op, itr::SkipOfType{T, <:AbstractArray}, ifirst::Integer, ilast::Integer,
+    blksize::Int
+) where T
+    A = itr.x
+    if ifirst == ilast
+        @inbounds a1 = A[ifirst]
+        if a1 isa T
+            return nothing
+        else
+            return Some(mapreduce_first(f, op, a1))
+        end
+    elseif ifirst + blksize > ilast
+        # sequential portion
+        local ai
+        i = ifirst
+        while i <= ilast
+            @inbounds ai = A[i]
+            ai isa T || break
+            i += 1
+        end
+        i > ilast && return nothing
+        a1 = ai::eltype(itr)
+        i += 1
+        while i <= ilast
+            @inbounds ai = A[i]
+            ai isa T || break
+            i += 1
+        end
+        i > ilast && return Some(mapreduce_first(f, op, a1))
+        a2 = ai::eltype(itr)
+        i += 1
+        v = op(f(a1), f(a2))
+        @simd for i = i:ilast
+            @inbounds ai = A[i]
+            if !(ai isa T)
+                v = op(v, f(ai))
+            end
+        end
+        return Some(v)
+    else
+        # pairwise portion
+        imid = (ifirst + ilast) >> 1
+        v1 = mapreduce_impl(f, op, itr, ifirst, imid, blksize)
+        v2 = mapreduce_impl(f, op, itr, imid+1, ilast, blksize)
+        if v1 === nothing && v2 === nothing
+            return nothing
+        elseif v1 === nothing
+            return v2
+        elseif v2 === nothing
+            return v1
+        else
+            return Some(op(something(v1), something(v2)))
+        end
+    end
+end
+
+union_poptype(::Type{T}, ::Type{Union{T,S}}) where {T, S} = S
+union_poptype(::Type{T}, ::Type{S}) where {T, S} = S
+union_poptype(::Type{T}, ::Type{T}) where {T} = Union{}
```

Comment:
This function is clever, but I noticed that this does not work in some generic cases. For example, you may want `skipoftype(Union{Missing,Nothing}, [1.0, missing, nothing])`.  However, this does not work ATM:

```julia
julia> union_poptype(Union{Missing, Nothing}, Union{Missing, Float64})
Union{Missing, Float64}

julia> Core.Compiler.typesubtract(Union{Missing, Float64}, Union{Missing, Nothing})
Float64
````

I guess one possibility is to use `Core.Compiler.typesubtract` but I suppose that'd break inference.

Reply:
I hadn't intended it to work with unions like that, but I'll see if I can think of a way to incorporate that.

Reply:
Having trouble thinking of something that works for that case, but it's not exported so restricting it to the use case I intended (subtract one type, not a union) seems reasonable to me.